src/improvements/template/OPCMUpgradeV200.sol

```diff
@@ -0,0 +1,98 @@
+    /// @return The address of the OPCM
+    function opcm() public pure override returns (address) {
+        return OPCM;
+    }
+
+    /// @notice Returns the storage write permissions
+    /// currently used OPCM is a dummy so no storage writes are needed
+    /// update this function to return all the storage writes permissions
+    /// required as per the OPCM contract.
+    /// @return Array of storage write permissions
+    function _taskStorageWrites() internal pure virtual override returns (string[] memory) {
+        string[] memory storageWrites = new string[](10);
+        storageWrites[0] = "OPContractsManager";
+        storageWrites[1] = "SuperchainConfig";
+        storageWrites[2] = "ProtocolVersions";
+        storageWrites[3] = "SystemConfigProxy";
+        storageWrites[4] = "L1CrossDomainMessengerProxy";
+        storageWrites[5] = "L1ERC721BridgeProxy";
+        storageWrites[6] = "L1StandardBridgeProxy";
+        storageWrites[7] = "DisputeGameFactory";
+        storageWrites[8] = "OptimismPortalProxy";
+        storageWrites[9] = "OptimismMintableERC20Factory";
+
+        return storageWrites;
+    }
+
+    /// @notice Sets up the template with prestate inputs from a TOML file
+    /// @param taskConfigFilePath Path to the TOML configuration file
+    function _templateSetup(string memory taskConfigFilePath) internal override {
+        OPCMUpgrade[] memory opcmUpgrade = abi.decode(
+            vm.parseToml(vm.readFile(taskConfigFilePath), ".opcmUpgrades.absolutePrestates"), (OPCMUpgrade[])
+        );
+
+        for (uint256 i = 0; i < opcmUpgrade.length; i++) {
+            opcmUpgrades[opcmUpgrade[i].chainId] = opcmUpgrade[i].absolutePrestate;
+        }
+    }
+
+    /// @notice build the task action for all l2chains in the task
+    /// in a single call to the OPCM.upgrade() function.
+    function _buildSingle() internal override {
+        AddressRegistry.ChainInfo[] memory chains = addrRegistry.getChains();
+        IOPContractsManager.OpChainConfig[] memory opChainConfigs =
+            new IOPContractsManager.OpChainConfig[](chains.length);
+
+        for (uint256 i = 0; i < chains.length; i++) {
+            opChainConfigs[i] = IOPContractsManager.OpChainConfig({
+                systemConfigProxy: ISystemConfig(addrRegistry.getAddress("SystemConfigProxy", chains[i].chainId)),
+                proxyAdmin: IProxyAdmin(addrRegistry.getAddress("ProxyAdmin", chains[i].chainId)),
+                absolutePrestate: opcmUpgrades[chains[i].chainId]
+            });
+        }
+
+        vm.label(opcm(), "OPCM");
+
+        (bool success,) = opcm().delegatecall(abi.encodeCall(IOPContractsManager.upgrade, (opChainConfigs)));
+        require(success, "OPCMUpgrateTemplate: failed to upgrade OPCM");
+    }
+
+    /// @notice validate the task for a given l2chain
+    /// for this dummy opcm there are no validations per l2 chain
+    /// for a real OPCM instance, add the validations per l2chain
```

Comment:
A lot of these comments can be updated. Also we will need to figure out what validations we want to perform. I _think_ we will want to use this: https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/interfaces/L1/IStandardValidator.sol